doc: present how to create a vim-plugin overlay

teto opened the pull request:
it is possible to maintain an out of tree list of neovim plugins with this method

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
doc: present how to create a vim-plugin overlay

it is possible to maintain an out of tree list of neovim plugins with this method

Update doc/languages-frameworks/vim.section.md

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented on doc/languages-frameworks/vim.section.md:
```diff
@@ -154,6 +154,22 @@ in
 
+You can use the updater script to generate basic packages out of a custom vim
+plugin list:
+`
+pkgs/misc/vim-plugins/update.py -i vim-plugin-names -o generated.nix --no-commit
+`
```
```suggestion
```
pkgs/misc/vim-plugins/update.py -i vim-plugin-names -o generated.nix --no-commit
```
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-do-i-add-a-vim-scripts-plugin-to-vim/14020/3

SuperSandro2000 commented on doc/languages-frameworks/vim.section.md:
````diff
@@ -154,6 +154,22 @@ in
 
+You can use the updater script to generate basic packages out of a custom vim
+plugin list:
+```
+pkgs/misc/vim-plugins/update.py -i vim-plugin-names -o generated.nix --no-commit
+```
+with the content of vim-plugin-names be like:
+```
+pwntester/octo.nvim
+```
````
```suggestion
```ShellSession
$ pkgs/misc/vim-plugins/update.py -i vim-plugin-names -o generated.nix --no-commit
```
with the content of vim-plugin-names be like:
```ShellSession
$ pwntester/octo.nvim
```
```

jlesquembre commented on doc/languages-frameworks/vim.section.md:
````diff
@@ -181,6 +181,22 @@ of precompiled grammars, you can use `nvim-treesitter.withPlugins` function:
+You can reference the generated vim plugins by:
+```nix
+myVimPlugins = pkgs.vimPlugins.extend (
+  (pkgs.callPackage generated.nix {})
````
It should be
```nix
    (pkgs.callPackage ./generated.nix {
      buildVimPluginFrom2Nix = pkgs.vimUtils.buildVimPluginFrom2Nix;
    })
```

teto replied:
fixed

teto commented:
Now that I've removed quite a bit of vim doc (related to VAM, the vim addon manager), I think there is space for this bit of information. Can we merge or are there still comments ?

jlesquembre commented:
Looks good to me, I'd like to get it finally merged

fricklerhandwerk commented on doc/languages-frameworks/vim.section.md:
````diff
@@ -182,6 +182,23 @@ of precompiled grammars, you can use `nvim-treesitter.withPlugins` function:
+```
+pkgs/applications/editors/vim/plugins/update.py -i vim-plugin-names -o generated.nix --no-commit
+```
+with the content of "vim-plugin-names" be like:
````
```suggestion
with the contents of `vim-plugin-names` being for example:
```

Also please use empty lines between paragraphs for better source readability.

Commit:
doc(neovim): remove the mention of plug for neovim

This was removed to simplify configuration. One could create a function that converts the plug format to vim native package format (only plugin system supported for neovim)  and upstream it to nixpkgs if that's an issue

Commit:
doc(vim): take into account plug non-support

Apply suggestions from code review

Co-authored-by: Valentin Gagarin <valentin.gagarin@tweag.io>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-patch-files-downloaded-by-vim-plugins/21248/5

teto commented:
I've added a mention about vim-plug being deprecated for neovim.

teto commented:
@jlesquembre @fricklerhandwerk would you mind doing another pass/merge if tha's fine ?

fricklerhandwerk commented on doc/languages-frameworks/vim.section.md:
````diff
@@ -243,10 +231,23 @@ Alternatively, set the number of processes to a lower count to avoid rate-limiti
+You can use the updater script to generate basic packages out of a custom vim
+plugin list:
+```
+pkgs/applications/editors/vim/plugins/update.py -i vim-plugin-names -o generated.nix --no-commit
````
Can you run this out of tree, i.e. with `nix-shell`?

teto replied:
That would be a logical followup indeed. Not sure how to expose that yet so I'll postponethe suggestion.

fricklerhandwerk commented on doc/languages-frameworks/vim.section.md:
````diff
@@ -243,10 +231,23 @@ Alternatively, set the number of processes to a lower count to avoid rate-limiti
+```
+pkgs/applications/editors/vim/plugins/update.py -i vim-plugin-names -o generated.nix --no-commit
+```
+with the contents of "vim-plugin-names" being for example:
````
```suggestion
with the contents of `vim-plugin-names` being for example:
```

fricklerhandwerk commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -148,7 +148,8 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
 
 - (Neo)Vim can not be configured with `configure.pathogen` anymore to reduce maintainance burden.
-Use `configure.packages` instead.
+Use `configure.packages` instead. Neovim can not be configured with plug anymore
+too (still works for vim).
```
```suggestion
  Use `configure.packages` instead.
- Neovim can not be configured with plug anymore (still works for vim).
```

fricklerhandwerk reviewed (changes requested):
This is a great addition, thank you. A few minor comments on presentation, other than that the explanations are very helpful. Please add newlines around markdown code blocks, this increases readability of the source.

A general thing which I would like to see, but would not block on:
Could we establish a more consistent naming of tools? Sometimes it is "Vim" or "Neovim", sometimes it is "vim" or "neovim".

On https://neovim.io they address it as "Vim" and "Neovim", and I think we should stick to that.

I know this is inconsistent in the context and in many areas of the manual, but we can only reasonably improve this one PR at a time.

teto commented:
@fricklerhandwerk  I've used Neovim everywhere as you mentioned and incorporated your good (=all) suggestions :) . Merging.
